#' Read VCF files into a GRangesList
#'
#' This function reads Variant Call Format (VCF) files into a GRanges object
#' and combines them in a GRangesList.  In addition to loading the files, this
#' function applies the same seqlevel style to the GRanges objects as the
#' reference genome passed in the 'genome' parameter.
#'
#' @param vcf_files Character vector of VCF file names
#' @param sample_names Character vector of sample names
#' @param genome A string matching the name of a BSgenome library
#'               corresponding to the reference genome of your VCFs
#' @param group Selector for a seqlevel group.  All seqlevels outside
#'              of this group will be removed.  Possible values:
#'              "all", "auto", "sex", "auto+sex" (default), "everything",
#'              and "circular".  See 'extractSeqlevelsByGroup' for more
#'              information.
#'
#' @return A GRangesList containing the GRanges obtained from 'vcf_files'
#'
#' @importFrom BiocGenerics match
#' @importFrom VariantAnnotation readVcf
#' @importFrom SummarizedExperiment rowRanges
#' @importFrom GenomeInfoDb "seqlevelsStyle<-"
#' @importFrom GenomeInfoDb "organism"
#' @importFrom parallel detectCores
#' @importFrom parallel mclapply
#'
#' @examples
#' # The example data set consists of three colon samples, three intestine
#' # samples and three liver samples.  So, to map each file to its appropriate
#' # sample name, we create a vector containing the sample names:
#' sample_names <- c ( "colon1", "colon2", "colon3",
#'                     "intestine1", "intestine2", "intestine3",
#'                     "liver1", "liver2", "liver3" )
#'
#' # We assemble a list of files we want to load.  These files match the
#' # sample names defined above.
#' vcf_files <- list.files(system.file("extdata", 
#'                                     package="MutationalPatterns"),
#'                                     pattern = ".vcf", full.names = TRUE)
#'
#' # Get a reference genome BSgenome object.
#' ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
#' library("BSgenome")
#' library(ref_genome, character.only = TRUE)
#'
#' # This function loads the files as GRanges objects
#' vcfs <- read_vcfs_as_granges(vcf_files, sample_names, ref_genome)
#'
#' @export

read_vcfs_as_granges <- function(vcf_files, sample_names, genome = "-",
                                 group = "auto+sex")
{
    # Check sample names
    if (length(vcf_files) != length(sample_names))
        stop("Please provide the same number of sample names as VCF files")

    # Check whether the user has adapted to the new behavior of the function.
    if (genome == "-")
        stop(paste("Please pass a reference genome string in the 'genome'",
                   "parameter.  This string can be obtained using",
                   "available.genomes() from the BSgenome package."))

    ref_genome <- base::get(genome)
    ref_organism <- GenomeInfoDb::organism(ref_genome)
    ref_style <- seqlevelsStyle(ref_genome)

    # Name the VCF's genome as the name of the genome build instead of
    # the BSgenome package name.
    genome_name <- genome(ref_genome)[[1]]

    # Check the class of the reference genome
    if (!(class(ref_genome) == "BSgenome"))
        stop("Please provide the name of a BSgenome object.")

    num_cores <- detectCores()

    # On confined OS environments, this value can be NA.
    # One core will be substracted from the total, so we
    # set this to 2.
    if (is.na(num_cores))
        num_cores = 1

    vcf_list <- GRangesList(mclapply (vcf_files, function (file)
    {
        # Use VariantAnnotation's readVcf, but only store the
        # GRanges information in memory.  This speeds up the
        # loading significantly.
        vcf <- rowRanges(readVcf (file, genome_name))

        # Convert to a single naming standard.
        seqlevelsStyle(vcf) <- ref_style[1]

        groups <- c()
        if (group != "everything")
        {
            if (group == "auto+sex")
            {
                groups <- c(extractSeqlevelsByGroup(species = ref_organism,
                                                    style = ref_style,
                                                    group = "auto"),
                            extractSeqlevelsByGroup(species = ref_organism,
                                                    style = ref_style,
                                                    group = "sex"))

                # In some cases, the seqlevelsStyle returns multiple styles.
                # In this case, we need to do a little more work to extract
                # a vector of seqlevels from it.
                groups_names <- names(groups)
                if (! is.null(groups_names))
                {
                    # The seqlevels in the groups are now duplicated.
                    # The following code deduplicates the list items, so that
                    # creating a data frame will work as expected.
                    unique_names <- unique(groups_names)
                    groups <- llply(unique_names, function(x) groups[groups_names == x])
                    groups <- llply(groups, unlist, recursive = F)

                    # In case there are multiple styles applied, we only use the first.
                    groups <- unique(as.vector(groups[[1]]))
                }
            }
            else
            {
                groups <- extractSeqlevelsByGroup ( species = ref_organism,
                                                   style = ref_style,
                                                   group = group )
                groups <- unique(as.vector(t(groups)))
            }

            vcf <- keepSeqlevels(vcf, groups)
        }

        # Find and exclude positions with indels or multiple
        # alternative alleles.
        rem <- which(all(!( !is.na(match(vcf$ALT, DNA_BASES)) &
                            !is.na(match(vcf$REF, DNA_BASES)) &
                            (lengths(vcf$ALT) == 1) )))

        if (length(rem) > 0)
        {
            vcf = vcf[-rem]
            warning(length(rem),
                    " position(s) with indels and multiple",
                    " alternative alleles are removed.")
        }

        return(vcf)
    }, mc.cores = num_cores))

    # Set the provided names for the samples.
    names(vcf_list) <- sample_names

    return(vcf_list)
}

##
## Deprecated variants
##

read_vcf <- function(vcf_files, sample_names, genome="-", style="UCSC")
{
    .Defunct("read_vcfs_as_granges", package="MutationalPatterns",
            msg=paste("This function has been removed.  Use",
                        "'read_vcfs_as_granges' instead.  The new function",
                        "automatically renames the seqlevel style for you,",
                        "so you no longer need to run 'rename_chrom' either."))
}

vcf_to_granges <- function(vcf_files, sample_names, genome="-", style="UCSC")
{
    # Show the same error message as 'read_vcf()'.
    read_vcf()
}

rename_chrom <- function(granges, style = "UCSC")
{
    .Defunct("rename_chrom", package="MutationalPatterns",
            msg = paste("This function has been removed."))
}
